test/box-luatest/gh_9997_update_splice_test.lua

```diff
@@ -0,0 +1,273 @@
+    local test_cases = generate_test_cases(tmpl)
+
+    for _, test_case in ipairs(test_cases) do
+        local ok, ret = pcall(tuple.update, tuple, test_case.ops)
+        ret = ok and ret or tostring(ret)
+        local res = {ok = ok, ret = ret}
+        local msg = get_test_case_message(test_case, res)
+        t.assert_equals(res, test_case.res, msg)
+        if ok then
+            t.assert_equals(field_type(ret[2]), cg.params.type, msg)
+        end
+    end
+end
+
+-- Test lots of various variants of splice (':') of memtx space update.
+g.test_memtx_update_splice = function(cg)
+    local tmpl = generate_tuple_template(cg.params.type)
+    local test_cases = generate_test_cases(tmpl)
+
+    cg.server:exec(function(data_type, tmpl, test_cases,
+                            get_test_case_message_source, field_type_source)
+        local get_test_case_message = loadstring(get_test_case_message_source)()
+        local field_type = loadstring(field_type_source)()
+
+        local s = box.schema.space.create('test', {engine = 'memtx'})
+        s:create_index('pk')
+
+        for _, test_case in ipairs(test_cases) do
+            local ops = test_case.ops
+            s:replace(tmpl)
+            local ok, ret = pcall(s.update, s, tmpl[1], ops)
+            ret = ok and ret or tostring(ret)
+            local res = {ok = ok, ret = ret}
+            local msg = get_test_case_message(test_case, res)
+            t.assert_equals(res, test_case.res, msg)
+            if ok then
+                t.assert_equals(field_type(ret[2]), data_type, msg)
+            end
+        end
+    end, {cg.params.type, tmpl, test_cases,
+          get_test_case_message_source, field_type_source})
+end
+
+-- Test lots of various variants of splice (':') of memtx space upsert.
+g.test_memtx_upsert_splice = function(cg)
+    local tmpl = generate_tuple_template(cg.params.type)
+    local test_cases = generate_test_cases(tmpl, true)
+
+    cg.server:exec(function(data_type, tmpl, test_cases,
+                            get_test_case_message_source, field_type_source)
+        local get_test_case_message = loadstring(get_test_case_message_source)()
+        local field_type = loadstring(field_type_source)()
+
+        local s = box.schema.space.create('test', {engine = 'memtx'})
+        s:create_index('pk')
+
+        for _, test_case in ipairs(test_cases) do
+            local ops = test_case.ops
+            s:replace(tmpl)
+            local ok, ret = pcall(s.upsert, s, tmpl, ops)
+            ret = ok and s:get(tmpl[1]) or tostring(ret)
+            local res = {ok = ok, ret = ret}
+            local msg = get_test_case_message(test_case, res)
+            t.assert_equals(res, test_case.res, msg)
+            if ok then
+                t.assert_equals(field_type(ret[2]), data_type, msg)
+            end
+        end
+    end, {cg.params.type, tmpl, test_cases,
+          get_test_case_message_source, field_type_source})
+end
```

Comment:
My overall impression is the test is too complicated to test the introduced changes. In includes fuzzy testing, helpers to improve report on failed test case as cases are run from a list, generating reference result instead of simply set it by hand. 

I guess making a fuzzy test for update/upsert is different task. Here we'd better make a simple functional test. Just a bunch of assertions is the best)